consensus/types/src/beacon_state/progressive_balances_cache.rs

```diff
@@ -0,0 +1,170 @@
+use crate::beacon_state::balance::Balance;
+use crate::{BeaconState, BeaconStateError, ChainSpec, Epoch, EthSpec};
+use arbitrary::Arbitrary;
+use safe_arith::SafeArith;
+use serde_derive::{Deserialize, Serialize};
+use strum::{Display, EnumString, EnumVariantNames};
+
+/// This cache keeps track of the accumulated target attestation balance for the current & previous
+/// epochs. The cached values can be utilised by fork choice to calculate unrealized justification
+/// and finalization instead of converting epoch participation arrays to balances for each block we
+/// process.
+#[derive(Default, Debug, PartialEq, Arbitrary, Clone)]
+pub struct ProgressiveBalancesCache {
+    inner: Option<Inner>,
+}
+
+#[derive(Debug, PartialEq, Arbitrary, Clone)]
+struct Inner {
+    pub current_epoch: Epoch,
+    pub previous_epoch_target_attesting_balance: Balance,
+    pub current_epoch_target_attesting_balance: Balance,
+}
+
+impl ProgressiveBalancesCache {
+    pub fn initialize(
+        &mut self,
+        current_epoch: Epoch,
+        previous_epoch_target_attesting_balance: Balance,
+        current_epoch_target_attesting_balance: Balance,
+    ) {
+        self.inner = Some(Inner {
+            current_epoch,
+            previous_epoch_target_attesting_balance,
+            current_epoch_target_attesting_balance,
+        });
+    }
+
+    pub fn is_initialized(&self) -> bool {
+        self.inner.is_some()
+    }
+
+    /// When a new target attestation has been processed, we update the cached
+    /// `current_epoch_target_attesting_balance` to include the validator effective balance.
+    /// If the epoch is neither the current epoch nor the previous epoch, an error is returned.
+    pub fn on_new_target_attestation(
```

Comment:
Discussed with @michaelsproul earlier on this - currently the bulk of cache update logic lives in state processing, which makes the cache simple, however this approach has some downside:
1. The correctness would depend largely on the calling code (e.g. only unslashed validator balance should be added), and it may not be obvious that a change in state processing can break the progressive cache. This makes the cache quite brittle.
2. This adds some additional logic to epoch/block processing, which is not ideal.

I'll look into moving some of the logic into the cache.

Reply:
Moved all cache mutation logic to this file in `state_processing`: 
https://github.com/sigp/lighthouse/blob/132bf8416500387d136c9caeb05a8479d4d1c536/consensus/state_processing/src/common/update_progressive_balances_cache.rs